language/diem-framework/modules/Diem.move

```diff
@@ -485,97 +551,338 @@ module Diem {
         Roles::assert_treasury_compliance(tc_account);
         assert(!is_synthetic_currency<CoinType>(), Errors::invalid_argument(EIS_SYNTHETIC_CURRENCY));
-        assert(!exists<Preburn<CoinType>>(Signer::address_of(account)), Errors::already_published(EPREBURN));
-        move_to(account, create_preburn<CoinType>(tc_account))
+        publish_preburn_queue<CoinType>(account)
     }
-    spec fun publish_preburn_to_account {
-        modifies global<Preburn<CoinType>>(Signer::spec_address_of(account));
+    spec fun publish_preburn_queue_to_account {
+        pragma opaque;
+        let account_addr = Signer::spec_address_of(account);
+        modifies global<PreburnQueue<CoinType>>(account_addr);
         /// The premission "PreburnCurrency" is granted to DesignatedDealer [[H4]][PERMISSION].
         /// Must abort if the account does not have the DesignatedDealer role.
         include Roles::AbortsIfNotDesignatedDealer;
-        /// Preburn is published under the DesignatedDealer account.
-        ensures exists<Preburn<CoinType>>(Signer::spec_address_of(account));
+        /// PreburnQueue is published under the DesignatedDealer account.
+        include PublishPreburnQueueAbortsIf<CoinType>;
+        include PublishPreburnQueueEnsures<CoinType>;
+        ensures exists<PreburnQueue<CoinType>>(account_addr);
 
         include Roles::AbortsIfNotTreasuryCompliance{account: tc_account};
         include AbortsIfNoCurrency<CoinType>;
         aborts_if is_synthetic_currency<CoinType>() with Errors::INVALID_ARGUMENT;
-        aborts_if exists<Preburn<CoinType>>(Signer::spec_address_of(account)) with Errors::ALREADY_PUBLISHED;
+        aborts_if exists<PreburnQueue<CoinType>>(account_addr) with Errors::ALREADY_PUBLISHED;
+        aborts_if exists<Preburn<CoinType>>(account_addr) with Errors::INVALID_STATE;
 
     }
 
     ///////////////////////////////////////////////////////////////////////////
 
+
+    /// Upgrade a designated dealer account from using a single `Preburn`
+    /// resource to using a `PreburnQueue` resource so that multiple preburn
+    /// requests can be outstanding in the same currency for a designated dealer.
+    fun upgrade_preburn<CoinType>(account: &signer)
+    acquires Preburn {
+        Roles::assert_designated_dealer(account);
+        let sender = Signer::address_of(account);
+        let preburn_exists = exists<Preburn<CoinType>>(sender);
+        let preburn_queue_exists = exists<PreburnQueue<CoinType>>(sender);
+        // The DD must already have an existing `Preburn` resource, and not a
+        // `PreburnQueue` resource already, in order to be upgraded.
+        if (preburn_exists && !preburn_queue_exists) {
+            let Preburn { to_burn } = move_from<Preburn<CoinType>>(sender);
+            destroy_zero(to_burn);
+            publish_preburn_queue<CoinType>(account);
+        } else {
+            assert(preburn_queue_exists && !preburn_exists, Errors::invalid_state(EPREBURN_QUEUE))
```

Comment:
I really wish we had a `prover_assert` or `unreachable` construct to instruct the prover to ensure that this cannot happen... In the absence of this, could we add a comment?

Reply:
Alternatively, you could gate the call to `upgrade_preburn` with the check above and avoid introducing/propagating this error code.

Reply:
Actually, I think you can do this with `spec { assert preburn_queue_exists && !preburn_exists }`

Reply:
I'll give this a shot! I hope this works :)

Reply:
Had a look and I actually wound up being able to remove this assertion entirely by tightening one of the other specs. So even better :)